Handle "partial success" exit codes

markhepburn opened the pull request:
Grep, ripgrep, probably others will use an exit status of 2 if there
was output, but a non-fatal error was encountered (for example,
attempting to access a file).

See BurntSushi/ripgrep#1278 for the motivation behind this one.  I'm using helm-ag with ripgrep, and there's a corner-case on windows I'm encountering where certain inaccessible files are being searched, resulting in an error.  This means that there is the expected search output, but because ripgrep (now) respects grep exit-code behaviour it returns 2, and helm-ag therefore tells you there was no output.

I can `(setq helm-ag-base-command "rg --no-messages --smart-case --no-heading --line-number")` (the `--no-messages` argument means no "soft" error messages will be output, just the search results), but the exit status still needs to be handled by helm-ag.

syohex commented:
Sorry very late reply. `ag` silver searcher uses exit code 2 as error code. So I suppose that like following patch is better than this PR.

```diff
diff --git a/helm-ag.el b/helm-ag.el
index 58a5c17..f266489 100644
--- a/helm-ag.el
+++ b/helm-ag.el
@@ -264,6 +264,10 @@ Default behaviour shows finish and result in mode-line."
       (while (re-search-forward "^\\([^:]+\\)" nil t)
         (replace-match (abbreviate-file-name (match-string-no-properties 1)))))))
 
+(defun helm-ag--command-succeeded-p (cmd exit-status)
+  (cond ((string= cmd "rg") (member exit-status '(0 2))) ;; add more commands if necessary
+        (t (zerop exit-status))))
+
 (defun helm-ag--init ()
   (let ((buf-coding buffer-file-coding-system))
     (helm-attrset 'recenter t)
@@ -278,9 +282,9 @@ Default behaviour shows finish and result in mode-line."
         (let ((ret (apply #'process-file (car cmds) nil t nil (cdr cmds))))
           (if (zerop (length (buffer-string)))
               (error "No ag output: '%s'" helm-ag--last-query)
-            (unless (memq ret '(0 2))
+            (unless (helm-ag--command-succeeded-p (car cmds) ret)
               (unless (executable-find (car cmds))
-                (error "'ag' is not installed."))
+                (error "'%s' is not installed." (car cmds)))
               (error "Failed: '%s'" helm-ag--last-query))))
         (when helm-ag--buffer-search
           (helm-ag--abbreviate-file-name))
```

markhepburn commented:
Looks good to me :)  The only thing that springs to mind is wondering if you'd want some extension point in `helm-ag--command-succeeded-p` as well, in case someone uses a third command.

syohex commented:
Please update PR if you have no problem about my above patch style ?

markhepburn commented:
Sorry for my even later reply!!  I added a patch from your suggestion.  I've also added one on top that shows what I was talking about; rather than hard-code the rg case, let the user optionally specify the exit code(s).  I'm happy to revert it though if you think that's too much.

jcs090218 commented:
I feel like this is good enough! Thanks for the contribution and the patch! :)

Merged!
